src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/FrozenArrayNode.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using Internal.Text;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents a frozen array with contents preinitialized from another RVA static field
+    /// </summary>
+    public class FrozenStaticFieldArrayNode : EmbeddedObjectNode, ISymbolDefinitionNode
+    {
+        private FieldDesc _arrayField;
+        private int _pointerSize; 
+
+        public FrozenStaticFieldArrayNode(FieldDesc arrayField, TargetDetails target)
+        {
+            _arrayField = arrayField;
+            _pointerSize = target.PointerSize;
+        }
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__FrozenArr_").Append(nameMangler.GetMangledFieldName(_arrayField));
+        }
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        int ISymbolNode.Offset => 0;
+
+        int ISymbolDefinitionNode.Offset
+        {
+            get
+            {
+                // The frozen array symbol points at the EEType portion of the object, skipping over the sync block
+                return OffsetFromBeginningOfArray + _pointerSize;
+            }
+        }
+
+        private IEETypeNode GetEETypeNode(NodeFactory factory)
+        {
+            return factory.GetLocalTypeSymbol(_arrayField.PreInitDataField.FieldType);
+        }
+
+        public override void EncodeData(ref ObjectDataBuilder dataBuilder, NodeFactory factory, bool relocsOnly)
+        {           
+            // Sync Block
+            dataBuilder.EmitZeroPointer();
+
+            // EEType
+            dataBuilder.EmitPointerReloc(GetEETypeNode(factory));
+          
+            var arrType = _arrayField.FieldType as ArrayType;
+            if (arrType == null || !arrType.IsSzArray)
+            {
+                // We only support single dimensional arrays
+                throw new NotSupportedException();
+            }
+
+            FieldDesc arrayDataField = _arrayField.PreInitDataField;
```

Comment:
Encoding of data should be flexible to allow the backend compiler to compute preinitialized data as well as accepting input from the IL dll that describes the preinitialized data.